ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PullFilterNode.java

```diff
@@ -0,0 +1,823 @@
+    final WindowBounds windowBounds = new WindowBounds();
+
+    new WindowBoundsExtractor().process(rewrittenPredicate, windowBounds);
+    return windowBounds;
+  }
+
+  /**
+   * Validate the WHERE clause for pull queries.
+   * 1. There must be exactly one equality condition per key
+   * or one IN predicate that involves a key.
+   * 2. An IN predicate can refer to a single key.
+   * 3. The IN predicate cannot be combined with other conditions.
+   * 4. Only AND is allowed.
+   * 5. If there is a multi-key, conditions on all keys must be specified.
+   * 6. The IN predicate cannot use multi-keys.
+   */
+  private final class Validator extends TraversalExpressionVisitor<Object> {
+
+    @Override
+    public Void process(final Expression node, final Object context) {
+      if (!(node instanceof  LogicalBinaryExpression)
+          && !(node instanceof  ComparisonExpression)
+          && !(node instanceof  InPredicate)) {
+        throw invalidWhereClauseException("Unsupported expression in WHERE clause: " + node, false);
+      }
+      super.process(node, context);
+      return null;
+    }
+
+    @Override
+    public Void visitLogicalBinaryExpression(
+        final LogicalBinaryExpression node,
+        final Object context
+    ) {
+      if (node.getType() != LogicalBinaryExpression.Type.AND) {
+        throw invalidWhereClauseException("Only AND expressions are supported: " + node, false);
+      }
+      process(node.getLeft(), context);
+      process(node.getRight(), context);
+      return null;
+    }
+
+    @Override
+    public Void visitComparisonExpression(
+        final ComparisonExpression node,
+        final Object context
+    ) {
+      final UnqualifiedColumnReferenceExp column = getColumnRefSide(node);
+
+      final ColumnName columnName = column.getColumnName();
+      if (columnName.equals(SystemColumns.WINDOWSTART_NAME)
+          || columnName.equals(SystemColumns.WINDOWEND_NAME)) {
+        final Type type = node.getType();
+        if (!VALID_WINDOW_BOUND_COMPARISONS.contains(type)) {
+          throw invalidWhereClauseException(
+              "Unsupported " + columnName + " bounds: " + type, true);
+        }
+        if (!isWindowed) {
+          throw invalidWhereClauseException(
+              "Cannot use WINDOWSTART/WINDOWEND on non-windowed source",
+              false);
+        }
+        return null;
+      } else {
+        final Column col = schema.findColumn(columnName)
+            .orElseThrow(() -> invalidWhereClauseException(
+                "Bound on non-key column " + columnName, isWindowed));
+
+        if (col.namespace() == Namespace.KEY) {
+          if (node.getType() != Type.EQUAL) {
+            throw invalidWhereClauseException(
+                "Bound on key columns '" + getSource().getSchema().key()
+                    + "' must currently be '='",
+                isWindowed);
+          }
+          if (containsINkeys || seenKeys.get(col.index())) {
```

Comment:
Currently, you have to check for this case in both `visitComparisonExpression` and  `visitInPredicate` since you don't know whether you're going to hit a comparison or in predicate first.

If you set `containsINkeys` and `seenKeys` in `ExtractKeyValues`, then you can do the check just once in one spot.

Reply:
I am not sure I follow what you mean. The validation happens first before we extract keys so how can I set it in the `ExtractKeyValues` ? It makes sense to validate first so that we fail fast.

Reply:
I just meant that you don't know which expression you'll hit first. so you have to put the check twice whereas if you factored out the key use metrics, you could check for the error once.

If the ordering of things doesn't allow for this, then that's fine.